### PR TITLE
Implement error reporting for the new compiler (closes #305)

### LINKED ISSUE #305: Implement error reporting for the new compiler
- [x] Implement source-frame package
- [x] Refactor TestCafe errors
- [x] File reading error
- [x] Stack clean-ups (flagged)
- [x] Compilation error (require, babel, runtime error) (perform stack clean up, show stack in CLI)
- [x] Globals errors (incorrect args, etc.) (record callsites, show stack in CLI)
- [x] CLI support

### PR DESCRIPTION
**Pull request by inikulin**
\cc @churkin @kirovboris @AndreyBelym 
@VasilyStrelyaev please, check strings, comments, etc.

## Here is how it looks like in CLI
 - Clean stacks (no babel, node or testcafe stuff, just test sources)
 - Proper formatting
 - Fancy call sites for our API (we will use same appearance in reports)

**Our API error:**
image

**Babel/syntax error:**
image

**Runtime error:**
image

### CODE REVIEW

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

File: src/errors/index.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+export class GlobalsAPIError extends Error {
+    constructor (methodName, typeName, template, ...args) {
+        super(getText(template, ...args));
+
+        this.callsiteRecord = getCallsite(methodName, typeName);
+        this.constructor    = GlobalsAPIError;
+
+        // HACK: prototype properties doesn't work with the builtin subclasses
```
Comment (VasilyStrelyaev) [inline src/errors/index.js]:
```
        // HACK: prototype properties don't work with built-in subclasses
```

Comment (inikulin) [inline src/errors/index.js]: [Wiki](https://en.wikipedia.org/wiki/Built-in) says that `builtin` is valid

File: src/errors/message.js [hunk 2]
```diff
@@ -11,10 +11,11 @@ export const MESSAGE = {
     portNumberIsNotInteger:             'A port number should be a valid integer.',
     portsOptionRequiresTwoNumbers:      'The "--ports" option requires two numbers to be specified.',
     portIsNotFree:                      'The specified {portNum} port is already in use by another program.',
-    invalidHostname:                    'The specified "{hostname}" hostname cannot be resolved to the current machine.'
+    invalidHostname:                    'The specified "{hostname}" hostname cannot be resolved to the current machine.',
+    cantFindSpecifiedTestSource:        'Cannot find specified test source file at "{path}".',
```
Comment (VasilyStrelyaev) [inline src/errors/message.js]: just thought that we don't specify a test source file - we specify the path
that's why we can just say
```
    cantFindSpecifiedTestSource:        'Cannot find a test source file at "{path}".',
```
can't we?

Comment (inikulin) [inline src/errors/message.js]: Yep

File: test/server/compiler-test.js [hunk 3]
```diff
@@ -1,13 +1,21 @@
...
+var stripAnsi   = require('strip-ansi');
+var Compiler    = require('../../lib/compiler');
+var Role        = require('../../lib/api/common/role');
+var Hybrid      = require('../../lib/api/common/hybrid');
 
 describe('Compiler', function () {
-    this.timeout(10000);
+    this.timeout(20000);
+
+    // FIXME: Babel errors always contains POSIX-format file paths.
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
    // FIXME: Babel errors always contain POSIX-format file paths.
```

File: test/server/compiler-test.js [hunk 4]
```diff
@@ -38,6 +46,53 @@ describe('Compiler', function () {
...
+    function assertGlobalsAPIError (err, expected) {
+        assertError(err, expected);
+
+        expect(err.stack.indexOf(expected.message + '\n\n' + expected.callsite)).eql(0);
+        expect(stripAnsi(err.coloredStack)).eql(err.stack);
+    }
+
+    function assertStack (err, expected) {
+        // HACK: stackParser can't handle empty stacks correctly
+        // (it treats error message as stack frames).
```
Comment (VasilyStrelyaev) [inline test/server/compiler-test.js]:
```
        // (it treats error messages as stack frames).
```

File: test/server/data/test-suites/fixture-name-is-not-a-string/testfile.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// NOTE: we need to workaround compiler's test file assertion
+// (it expects at least one fixture definition with the
+// string argument to treat file as a test).
```
Comment (VasilyStrelyaev) [inline test/server/data/test-suites/fixture-name-is-not-a-string/testfile.js]:
```
// NOTE: we need to workaround the compiler's test file assertion
// (to treat a file as a test, it requires at least one fixture definition
//  with the string argument).
```

File: test/server/data/test-suites/test-name-is-not-a-string/testfile.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+fixture `Test name is not a string`;
+
+// NOTE: we need to workaround compiler's test file assertion
+// (it expects at least one test definition with the
+// string argument to treat file as a test).
```
Comment (VasilyStrelyaev) [inline test/server/data/test-suites/test-name-is-not-a-string/testfile.js]:
```
// NOTE: we need to workaround the compiler's test file assertion
// (to treat a file as a test, it requires at least one fixture definition
//  with the string argument).
```

File: src/api/globals/index.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+export default class Globals {
+    constructor (filename) {
+        this.filename       = filename;
+        this.currentFixture = null;
+        this.collectedTests = [];
+
+        var globals = this;
+
+        // NOTE: use named functions so they appear in the stacks
```
Comment (VasilyStrelyaev) [inline src/api/globals/index.js]:
```
        // NOTE: use named functions so that they appear in the stacks
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by kirovboris** [comment]
lgtm

**Comment by AndreyBelym** [comment]
lgtm

**Comment by churkin** [comment]
lgtm, @testcafe-build-bot \retest

**Comment by churkin** [comment]
@testcafe-build-bot \retest

**Pull request merged**